python_example.py

```diff
@@ -32,33 +44,32 @@ class URL:
   def __init__(self, looker, user, session_length, embed_url, force_logout_login=False):
     self.looker = looker
     self.user = user
-    self.path = '/login/embed/' + urllib.quote_plus(embed_url)
+    self.path = '/login/embed/' + urllib.parse.quote_plus(embed_url)
     self.session_length = json.dumps(session_length)
     self.force_logout_login = json.dumps(force_logout_login)
 
   def set_time(self):
     self.time = json.dumps(int(time.time()))
 
   def set_nonce(self):
-    self.nonce = json.dumps(binascii.hexlify(os.urandom(16)))
+    self.nonce = json.dumps(to_ascii(binascii.hexlify(os.urandom(16))))
 
   def sign(self):
     #  Do not change the order of these
-    string_to_sign = ""
-    string_to_sign = string_to_sign + self.looker.host           + "\n"
-    string_to_sign = string_to_sign + self.path                  + "\n"
-    string_to_sign = string_to_sign + self.nonce                 + "\n"
-    string_to_sign = string_to_sign + self.time                  + "\n"
-    string_to_sign = string_to_sign + self.session_length        + "\n"
-    string_to_sign = string_to_sign + self.user.external_user_id + "\n"
-    string_to_sign = string_to_sign + self.user.permissions      + "\n"
-    string_to_sign = string_to_sign + self.user.models           + "\n"
-    string_to_sign = string_to_sign + self.user.group_ids        + "\n"
-    string_to_sign = string_to_sign + self.user.external_group_id + "\n"
-    string_to_sign = string_to_sign + self.user.user_attributes  + "\n"
-    string_to_sign = string_to_sign + self.user.access_filters
-
-    signer = hmac.new(self.looker.secret, string_to_sign.encode('utf8'), sha1)
+    string_to_sign = "\n".join([self.looker.host,
+                                self.path,
+                                self.nonce,
+                                self.time,
+                                self.session_length,
+                                self.user.external_user_id,
+                                self.user.permissions,
+                                self.user.models,
+                                self.user.group_ids,
+                                self.user.external_group_id,
+                                self.user.user_attributes,
+                                self.user.access_filters])
```

Comment:
I trust this works the same as the old code if any of these are empty string? And, any type conversions to string still work the same way? The *seems* like a nice refactoring. But, as a non-python-expert is not not 100% obvious that it is functionally identical to the previous code.

Reply:
@jbandhauer sounds like I should further modify the python SSO example to omit parameters that are "empty" from being encoded in the argument, like you do with Ruby. I'll do that Monday. Thanks for the suggestion. I'll add you to the review for that PR, even though it looks like you'll catch it anyway ;)

Reply:
you could write this as:
```python
string_to_sign = "\n".join(filter(None, [self.looker.host,
                                         self.path,
                                         self.nonce,
                                         self.time,
                                         self.session_length,
                                         self.user.external_user_id,
                                         self.user.permissions,
                                         self.user.models,
                                         self.user.group_ids,
                                         self.user.external_group_id,
                                         self.user.user_attributes,
                                         self.user.access_filters]))
```